camel/o1datagen/o1datagen.py

```diff
@@ -0,0 +1,365 @@
+        is_correct = response.msgs[0].content.strip().lower() == "true"
+        logger.info("Answer verification result: %s", is_correct)
+        return is_correct
+
+    def monte_carlo_tree_search(
+        self, question: str, partial_solution: str = ""
+    ) -> tuple[str, bool]:
+        r"""Generate and verify answers using Monte Carlo Tree Search.
+
+        This method implements a Monte Carlo Tree Search approach
+        to find the best solution by iteratively generating answers
+        and scoring them against the golden answer.
+
+        Args:
+            question (str): The problem or question to be solved.
+            partial_solution (str, optional): A partial solution to build
+                upon. This canbe used to guide the search process with
+                existing progress. (default::obj:`""`)
+
+        Returns:
+            tuple[str, bool]: A tuple containing:
+                - str: The best solution found
+                (or empty string if no solution found)
+                - bool: True if a correct solution was found, False otherwise
+        """
+        logger.info("Starting Monte Carlo Tree Search")
+        best_solution = ""  # Initialize as empty string instead of None
+        best_score: float = 0.0
+        for i in range(self.search_limit):
+            # Generate new answer
+            current_solution = self.get_answer(question, partial_solution)
+            # Verify answer
+            is_correct = self.verify_answer(question, current_solution)
+            if is_correct:
+                logger.info("Correct answer found! Stopping search")
+                return current_solution, True
+            prompt = (
+                f"Please evaluate this solution and "
+                f"give a score between 0-1:\n"
+                f"Question: {question}\n"
+                f"Solution: {current_solution}\n"
+                f"Correct answer: {self.golden_answers.get(question, '')}\n"
+                f"Return a JSON object with a single field 'score' containing "
+                f"a float between 0 and 1, like this: {{'score': 0.85}}\n"
+            )
+            self.chat_agent.reset()
+            response = self.chat_agent.step(prompt)
+            try:
+                response_json = response.msgs[0].content.strip()
+                agent_response = AgentResponse.parse_raw(response_json)
+                score = agent_response.score
+                if score > best_score:
+                    best_score = score
+                    best_solution = current_solution
+                    # Exit early if we find a very good solution (score > 0.9)
+                    if score > 0.9:
+                        logger.info(
+                            "Found excellent solution with score %.2f. "
+                            "Stopping search early.",
+                            score,
+                        )
+                        return best_solution, False
```

Comment:
should we move this if statement out of  `if score > best_score`?

Reply:
okok